From: Ilija Tovilo <ilija.tovilo@me.com>
Date: Tue, 7 Jul 2026 20:32:58 +0200
Subject: Fix leak on double DatePeriod::__construct() call

Closes GH-22643

(cherry picked from commit 2d86f8cf489ea5410edd19d301c1aefe330599ab)
(cherry picked from commit 591e613e1cf6cc10d84df2c799ede593b7371bc1)
---
 NEWS                                                |  5 +++++
 ext/date/php_date.c                                 | 21 ++++++++++++++++++++-
 .../tests/DatePeriod_double_constructor_call.phpt   | 14 ++++++++++++++
 3 files changed, 39 insertions(+), 1 deletion(-)
 create mode 100644 ext/date/tests/DatePeriod_double_constructor_call.phpt

diff --git a/NEWS b/NEWS
index e3cb991..b7b43af 100644
--- a/NEWS
+++ b/NEWS
@@ -1,6 +1,11 @@
 PHP                                                                        NEWS
 |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
 
+Backported from 8.2.33
+
+- Date:
+  . Fixed leak on double DatePeriod::__construct() call. (ilutov)
+
 Backported from 8.2.32
 
 - OpenSSL:
diff --git a/ext/date/php_date.c b/ext/date/php_date.c
index 55bf2a0..954fa2e 100644
--- a/ext/date/php_date.c
+++ b/ext/date/php_date.c
@@ -4202,6 +4202,23 @@ static int date_period_initialize(timelib_time **st, timelib_time **et, timelib_
 	return retval;
 } /* }}} */
 
+static void date_period_reset(php_period_obj *period_obj)
+{
+	if (period_obj->start) {
+		timelib_time_dtor(period_obj->start);
+	}
+	if (period_obj->current) {
+		timelib_time_dtor(period_obj->current);
+	}
+	if (period_obj->end) {
+		timelib_time_dtor(period_obj->end);
+	}
+	if (period_obj->interval) {
+		timelib_rel_time_dtor(period_obj->interval);
+	}
+	memset(period_obj, 0, XtOffsetOf(php_period_obj, std));
+}
+
 /* {{{ Creates new DatePeriod object. */
 PHP_METHOD(DatePeriod, __construct)
 {
@@ -4224,7 +4241,7 @@ PHP_METHOD(DatePeriod, __construct)
 	}
 
 	dpobj = Z_PHPPERIOD_P(ZEND_THIS);
-	dpobj->current = NULL;
+	date_period_reset(dpobj);
 
 	if (isostr) {
 		zend_replace_error_handling(EH_THROW, NULL, &error_handling);
@@ -4261,6 +4278,8 @@ PHP_METHOD(DatePeriod, __construct)
 		}
 		dpobj->start_ce = date_ce_date;
 	} else {
+		DATE_CHECK_INITIALIZED(Z_PHPINTERVAL_P(interval)->initialized, Z_OBJCE_P(interval));
+
 		/* init */
 		php_interval_obj *intobj = Z_PHPINTERVAL_P(interval);
 
diff --git a/ext/date/tests/DatePeriod_double_constructor_call.phpt b/ext/date/tests/DatePeriod_double_constructor_call.phpt
new file mode 100644
index 0000000..551d272
--- /dev/null
+++ b/ext/date/tests/DatePeriod_double_constructor_call.phpt
@@ -0,0 +1,14 @@
+--TEST--
+Double DatePeriod::__construct() call
+--FILE--
+<?php
+
+$start = new \DateTime();
+$interval = new \DateInterval('P1D');
+$period = new \DatePeriod($start, $interval, 1);
+$period->__construct($start, $interval, 1);
+
+?>
+===DONE===
+--EXPECT--
+===DONE===
